### PR TITLE
Abstract shell

### PR DESCRIPTION
**Pull request by baptistemesta**
Hi,
I coded that And I think it would be useful for others to have it.

It's just a small framework to make development of shell easier.
There is an abstract shell that clients can extends and on which Commands can be contributed.
Each commands define their completers, name and business code.
Clients can define a context that will be given to commands when executed.

The CommandArgumentsCompleter allows to complete command name and then delegate the completion to completers of each command.

Basically to write your own shell you just have to extends BaseShell.
The most important thing here is the initShellCommands that must returns a list of ShellCommand.

If you find this interesting tell me.
Also there is no test yet but I can add some.

### CODE REVIEW

**Comment by adzubla** [comment]
It looks like many people had the same idea...

I evaluated CRaSH, clam-shell, spring-shell and felix but found that they are too much for my taste!

So I started my own little shell framework, with the idea to keep it simple, extensible and with minimal dependencies (you can use it without jline if you want).

You can have a look at it: https://github.com/adzubla/josh

**Comment by nikolavp** [comment]
I did almost the same. Look at mine https://github.com/nikolavp/commander-shell. I also had the requirement for exposing the shell through sshd. Will look into yours for ideas later :) :+1:
